lib/src/http.dart

```diff
@@ -345,56 +189,85 @@ Never invalidServerResponse(http.Response response) =>
-  final Pool _pool;
-  final http.Client _inner;
-
-  /// Creates a new client that allows no more than [maxActiveRequests]
-  /// concurrent requests.
-  ///
-  /// If [inner] is passed, it's used as the inner client for sending HTTP
-  /// requests. It defaults to `new http.Client()`.
-  _ThrottleClient(int maxActiveRequests, this._inner)
-      : _pool = Pool(maxActiveRequests);
+/// If there is an HTTP-related exception, an intermittent HTTP error response,
+/// or an async timeout, [fn] is run repeatedly until there is a successful
+/// response or at most seven total attempts have been made. If all attempts
+/// fail, the final exception is re-thrown.
+///
+/// Each attempt is run within a [Pool] configured with 16 maximum resources.
+Future<T> retryForHttp<T>(String operation, FutureOr<T> Function() fn) async {
+  return await retry(
+      () async => await _httpPool.withResource(() async => await fn()),
+      retryIf: (e) async =>
+          (e is PubHttpException && e.isIntermittent) ||
+          e is TimeoutException ||
+          e is HttpException ||
+          e is TlsException ||
+          e is SocketException ||
+          e is WebSocketException,
+      onRetry: (exception, retryCount) async =>
+          log.io('Retry #${retryCount + 1} for $operation'),
+      maxAttempts: math.max(
+        1, // Having less than 1 attempt doesn't make sense.
+        int.tryParse(Platform.environment['PUB_MAX_HTTP_RETRIES'] ?? '') ?? 7,
+      ));
+}
 
-  @override
-  Future<http.StreamedResponse> send(http.BaseRequest request) async {
-    var resource = await _pool.request();
-
-    http.StreamedResponse response;
-    try {
-      response = await _inner.send(request);
-    } catch (_) {
-      resource.release();
-      rethrow;
+extension Throwing on http.BaseResponse {
+  /// Throws [PubHttpResponseException], calls [fail], or does nothing depending
+  /// on the status code.
+  ///
+  /// If the code is in the 200 range, nothing is done. If the code is 408, 429,
+  /// or in the 500 range, [PubHttpResponseException] is thrown with
+  /// "isIntermittent" set to `true`. Otherwise, [PubHttpResponseException] is
+  /// thrown with "isIntermittent" set to `false`.
+  void throwIfNotOk() {
+    if (statusCode >= 200 && statusCode <= 299) {
+      return;
+    } else if (statusCode == HttpStatus.notAcceptable &&
+        request?.headers['Accept'] == pubApiHeaders['Accept']) {
+      fail('Pub ${sdk.version} is incompatible with the current version of '
+          '${request?.url.host}.\n'
+          'Upgrade pub to the latest version and try again.');
+    } else if (statusCode >= 500 ||
+        statusCode == HttpStatus.requestTimeout ||
+        statusCode == HttpStatus.tooManyRequests) {
+      // Throw if the response indicates a server error or an intermittent
+      // client error, but mark it as intermittent so it can be retried.
+      throw PubHttpResponseException(this as http.Response,
```

Comment:
If you have to cast, maybe it's better to do `extension Throwing on http.Response` or, make `PubHttpResponseException` accept an `BaseResponse`

Reply:
Extending `http.Response` breaks the extension for the _download method, which uses `http.StreamedResponse`. Vice-versa for if we extend `http.StreamedResponse`, since the versions method uses `http.Response`.

---

#### The change to make `PubHttpResponseException` accept `http.BaseResponse` works, but causes build errors in lish.dart:

<img width="775" alt="Screen Shot 2022-11-03 at 5 10 14 PM" src="https://user-images.githubusercontent.com/159021/199858288-24ef33e4-376c-48f2-8eb0-c45102e3dfc4.png">

---

#### If we change `handleJsonError` to accept `http.BaseResponse`, we get some more errors:

<img width="776" alt="Screen Shot 2022-11-03 at 5 11 17 PM" src="https://user-images.githubusercontent.com/159021/199858373-16e1308b-fdad-48f7-9f70-b4c917b13dcc.png">

---

#### And then finally, if we change `parseJsonResponse` to accept `http.BaseResponse`:

<img width="776" alt="Screen Shot 2022-11-03 at 5 12 24 PM" src="https://user-images.githubusercontent.com/159021/199858428-668ae775-e8ba-4eb5-95f9-1bd3f10d34bc.png">

---

So, this was the only way I could come up with to make it all work. 😅

Reply:
Fair point, we can clean it up later.

Reply:
One option could be to create `PubHttpResponse(BaseResponse, List<int>? body);` so that it has an optional body.

But maybe this is better left for a follow up PR...

Reply:
I think we actually have: https://github.com/dart-lang/pub/issues/3130

For refactoring `handleJsonError`, I think if we're going to make better output -- that's also a great opportunity to re-think the API interface and use something like `BaseResponse` instead.

---------------------------

Actually, it might be cleaner to cast from `BaseResponse` to `Response` in `handleJsonError`, because -- if it's not a `Response` then it's same as not being able to read the body.

Reply:
The cast from `BaseResponse` to `Response` looks good!

I ended up having to implement that in this PR because the cast in `throwIfNotOk` was causing tests to fail.

https://github.com/dart-lang/pub/blob/7afe52cd1aa68c4546dcfe595a53d8890ed3b024/lib/src/http.dart#L192-L197